zkevm-circuits/src/evm_circuit/execution/codecopy.rs

```diff
@@ -0,0 +1,476 @@
+            cb,
+            opcode,
+            step_state_transition,
+            Some(memory_copier_gas.gas_cost()),
+        );
+
+        Self {
+            same_context,
+            code_offset,
+            code_size,
+            account,
+            dst_memory_addr,
+            memory_expansion,
+            memory_copier_gas,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction<F>,
+        call: &Call<F>,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        // 1. `dest_offset` is the bytes offset in the memory where we start to
+        // write.
+        // 2. `code_offset` is the bytes offset in the current
+        // context's code where we start to read.
+        // 3. `size` is the number of
+        // bytes to be read and written (0s to be copied for out of bounds).
+        let [dest_offset, code_offset, size] =
+            [0, 1, 2].map(|i| block.rws[step.rw_indices[i]].stack_value());
+
+        // assign the code offset memory address.
+        self.code_offset.assign(
+            region,
+            offset,
+            Some(
+                code_offset.to_le_bytes()[..N_BYTES_MEMORY_ADDRESS]
+                    .try_into()
+                    .unwrap(),
+            ),
+        )?;
+
+        let account = if call.is_root {
+            call.callee_address
+        } else {
+            unimplemented!("CODECOPY does not support internal calls yet");
+        };
+        self.account.assign(
+            region,
+            offset,
+            Some(
+                account
+                    .to_scalar()
+                    .expect("could not convert address to scalar field"),
+            ),
+        )?;
+
+        let code = block
+            .bytecodes
+            .iter()
+            .find(|b| {
+                RandomLinearCombination::random_linear_combine(
+                    b.hash.to_le_bytes(),
+                    block.randomness,
+                ) == call.opcode_source
+            })
+            .expect("could not find current environment's bytecode");
+        self.code_size
+            .assign(region, offset, Some(F::from(code.bytes.len() as u64)))?;
+
+        // assign the destination memory offset.
+        let memory_address =
+            self.dst_memory_addr
+                .assign(region, offset, dest_offset, size, block.randomness)?;
+
+        // assign to gadgets handling memory expansion cost and copying cost.
+        let (_, memory_expansion_cost) = self.memory_expansion.assign(
+            region,
+            offset,
+            step.memory_word_size(),
+            [memory_address],
+        )?;
+        self.memory_copier_gas.assign(
+            region,
+            offset,
+            size.as_u64(),
+            memory_expansion_cost as u64,
```

Comment:
memory_expansion_cost is already a u64, so you don't need to cast this. I don't know why clippy didn't say anything about this.